telegram/ext/filters.py

```diff
@@ -892,38 +893,169 @@ class user(BaseFilter):
+            :meth:`remove_user`. Only update the entire set by
+            ``filter.user_ids/usernames = new_set``, if you are entirely sure that it is not
+            causing race conditions, as this will complete replace the current set of allowed
+            users.
+
+        Attributes:
+            user_ids(set(:obj:`int`), optional): Which user ID(s) to allow through.
+            usernames(set(:obj:`str`), optional): Which username(s) (without leading '@') to allow
+                through.
+            allow_empty(:obj:`bool`, optional): Whether updates should be processed, if no user
+                is specified in :attr:`user_ids` and :attr:`usernames`.
         Args:
-            user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow through.
-            username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow through.
-                If username starts with '@' symbol, it will be ignored.
+            user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow
+                through.
+            username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow
+                through. Leading '@'s in usernames will be discarded.
+            allow_empty(:obj:`bool`, optional): Whether updates should be processed, if no user
+                is specified in :attr:`user_ids` and :attr:`usernames`. Defaults to :obj:`False`
 
         Raises:
             ValueError: If chat_id and username are both present, or neither is.
 
         """
 
-        def __init__(self, user_id=None, username=None):
-            if not (bool(user_id) ^ bool(username)):
-                raise ValueError('One and only one of user_id or username must be used')
-            if user_id is not None and isinstance(user_id, int):
-                self.user_ids = [user_id]
-            else:
-                self.user_ids = user_id
+        def __init__(self, user_id=None, username=None, allow_empty=False):
+            self.allow_empty = allow_empty
+            self.__lock = Lock()
+
+            self._user_ids = set()
+            self._usernames = set()
+
+            self._set_user_ids(user_id)
+            self._set_usernames(username)
+
+        @staticmethod
+        def _parse_user_id(user_id):
+            if user_id is None:
+                return set()
+            if isinstance(user_id, int):
+                return {user_id}
+            return set(user_id)
+
+        @staticmethod
+        def _parse_username(username):
             if username is None:
-                self.usernames = username
-            elif isinstance(username, string_types):
-                self.usernames = [username.replace('@', '')]
-            else:
-                self.usernames = [user.replace('@', '') for user in username]
+                return set()
+            if isinstance(username, str):
+                return {username.replace('@', '')}
```

Comment:
I've just realized that this will replace all @ from the username.
Maybe we had a (minor) bug until now...

What do you think about removing only a single leading @ ?

```python
if username.startswith('@'):
    username = username[1:]
```

Reply:
"Allowed are a-z, 0-9 and underscores." Still, that might change in the future, so I changed it and also added it to the test cases.